Feature/refactor no enable func

adriantuk opened the pull request:
Hi Jae, This is a new branch - as we discussed, it does not contain option to disable/enable Approov SDK due to the fact that SDK does not support it yet.

adriantuk committed:
refactor: extract decision making logic into ApproovInterceptorExtensions interface

- Add ApproovInterceptorExtensions interface with default status handling methods
- Extract TokenFetchStatus decision logic from ApproovService into reusable interface
- Add CustomApproovDecisionMaker as example implementation
- Simplify ApproovService by delegating status decisions to configurable decision maker
- Improve extensibility for custom token fetch status handling

adriantuk committed:
Make handleCustomJWTStatus() return token, and accept just TokenFetchResult

adriantuk committed:
add handleInterceptorActionBasedOnRequest() to allow more control if request should be modyfied by approov based on the request itself.

adriantuk committed:
refactor: update decison makers to use TokenFetchResult directly; few other tweaks

adriantuk committed:
feat: add Approov protection toggle and update client behavior based on its state

adriantuk committed:
Fix a build issue

adriantuk committed:
Removing the capability to enable or disable the Approov SDK. This is pending because the SDK does not fully support it yet.

jexh committed:
Add a bunch of changes to move this towards a neater solution.

jexh commented:
Added a bunch of changes. Adrian, please review. There are still some items remaining:
* Exceptions should have a code to indicate common cases without consumers needing to interpret strings
* The message signature type needs to be fixed
Please chat to me about this tomorrow

devenney commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java:
```diff
@@ -276,31 +276,54 @@ static synchronized String getBindingHeader() {
      * Gets the interceptor extensions callback handlers.
      *
      * @return the interceptor extensions callback handlers or null if none set
+     * @deprecated Use getApproovServiceMutator instead
```
```suggestion
     * @deprecated Use getServiceMutator instead
```

Alternatively, rename the function to match (but be aware that `setServiceMutator()` should be updated for consistency in this case).

devenney commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java:
```diff
@@ -276,31 +276,54 @@ static synchronized String getBindingHeader() {
+    }
+
+    /**
+     * @deprecated Use setApproovServiceMutator instead
```
```suggestion
     * @deprecated Use setServiceMutator instead
```

Alternatively, rename the function to match (but be aware that `getServiceMutator()` should be updated for consistency in this case).

devenney commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java:
```diff
@@ -923,36 +895,26 @@ public Response intercept(Chain chain) throws IOException {
-        // protected by Approov and therefore potential subject to a MitM
-        if ((approovResults.getStatus() != Approov.TokenFetchStatus.SUCCESS) &&
-                (approovResults.getStatus() != Approov.TokenFetchStatus.UNPROTECTED_URL))
+            // add trace id header too (if it is available)
```
Is there missing code here? The comment implies we are doing something, but this is the end of the conditional.

Alternatively,

```suggestion
            // TODO: add trace id header too (if it is available)
```

adriantuk replied:
As this feature is 3.5.2, I'll remove the comment

devenney commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovServiceMutator.java:
```diff
@@ -0,0 +1,311 @@
+     *                          action or throw an ApproovException encoding
+     *                          the cause of the failure
+     */
+    default void handleSecureStringResult(Approov.TokenFetchResult approovResults, String operation, String key) throws ApproovException {
```
I think this should be renamed for consistency. We have:

* `handlePrecheckResult` (handles `.precheck()`)
* `handleFetchTokenResult` (handles `.fetchToken()`)
* `handleSecureStringResult` (handles `.fetchSecureString()`)

Thus, it seems this should be:

```suggestion
    default void handleFetchSecureStringResult(Approov.TokenFetchResult approovResults, String operation, String key) throws ApproovException {
```

devenney commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovServiceMutator.java:
```diff
@@ -0,0 +1,311 @@
+     *                          action or throw an ApproovException encoding
+     *                          the cause of the failure
+     */
+    default void handleCustomJWTResult(Approov.TokenFetchResult approovResults) throws ApproovException {
```
I think this should be renamed for consistency. We have:

* `handlePrecheckResult` (handles `.precheck()`)
* `handleFetchTokenResult` (handles `.fetchToken()`)
* `handleSecureStringResult` (handles `.fetchSecureString()`)
* `handleCustomJWTResult` (handles `fetchCustomJWT()`)

Thus, it seems this should be:

```suggestion
    default void handleFetchCustomJWTResult(Approov.TokenFetchResult approovResults) throws ApproovException {
```

devenney reviewed (dismissed):
Good restructure. Comments are around consistency of function naming, and a typo or two in our deprecation notices.

adriantuk committed:
Fix few build issues; Import LinkedHashSet; Resolve Brendan comments; Adjusted names of the mutator default methods to match the action being performed.

adriantuk committed:
Created a solid foundation for error reporting through custom error codes. Keept backwards compatibility by adding general 9xxx error codes when using deprecated constructor.
 - Introduced centralised error catalog so every Approov exception maps to named constant with a readable description
 - Made ApproovException (and subclasses) carry an explicit error code, expose getErrorCode() / getErrorDescription(), and embed the code in messages
 - Updated all Approov service, mutator, and custom decision-maker flows to supply the appropriate constants—covering SDK misuses, rejection cases, and retry-able network conditions so failures are now categorised consistently.

jexh committed:
Modify message signature to use a string (as required by the spec) instead of a byte sequence

adriantuk committed:
Revert "Created a solid foundation for error reporting through custom error codes. Keept backwards compatibility by adding general 9xxx error codes when using deprecated constructor."

This reverts commit 849e4885ab6673cf506a7dd0560f8173ff5c5386.

adriantuk committed:
Introduced the concept of ErrorCode. Refactored ApproovException and its subclasses to use specific error codes for improved error handling. Updated remaining classes to adopt the error code implementation. Deprecated constructors that do not utilize error codes.

adriantuk committed:
Fix build issue

richardmtaylor commented:
@adriantuk I assume we are still keeping the IOException hierarchy as well as adding these subcodes, since the documentation uses this already and is actually quite a neat solutiion for the basic error captures on prefetches etc,

adriantuk commented:
> @adriantuk I assume we are still keeping the IOException hierarchy as well as adding these subcodes, since the documentation uses this already and is actually quite a neat solutiion for the basic error captures on prefetches etc,

That's correct, ApproovException class still extends IOException, and this won't be changed - all exceptions are being  passed to parent class using ```super(message, cause);```

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovException.java:
```diff
@@ -17,17 +17,141 @@
+            return ERROR_UNKNOWN;
+        }
+        switch (status) {
+            case REJECTED:
```
I know it doesn't make sense for our code but I think we should make sure every status code is here - even SUCCESS.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovException.java:
```diff
@@ -17,17 +17,141 @@
+     * @param message is the basic information about the exception cause
+     * @param cause is the underlying cause of the exception
+     */
+    public ApproovException(Approov.TokenFetchStatus status, String message, Throwable cause) {
```
Is this one used? It doesn't seem like it would ever be possible because either you get the status  or an exception not both. I think you should remove it.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovNetworkException.java:
```diff
@@ -27,6 +29,26 @@ public class ApproovNetworkException extends ApproovException {
      */
     public ApproovNetworkException(String message) {
-        super(message);
+        super(ApproovException.ERROR_UNKNOWN, message);
```
Shouldn't this use ERROR_LEGACY like ApproovException?

jexh committed:
omment fix

jexh committed:
Review ApproovException error codes

adriantuk committed:
Remove error codes; introduce TokenFetchStatusException; deprecate ApproovNetworkException.

adriantuk committed:
simplify exception message handling

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovException.java:
```diff
@@ -19,15 +19,39 @@
+     * @param cause underlying cause of the exception
+     */
+    public ApproovException(Throwable cause) {
+        super(cause);
```
You should add a message indicating that a cause is wrapped.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovNetworkException.java:
```diff
@@ -9,24 +9,40 @@
+
+/**
+ * @deprecated Use {@link TokenFetchStatusException} instead. This subtype is retained only to avoid
+ *             breaking existing call sites; migrate any explicit catches to the base exception.
```
```suggestion
 *             breaking existing handlers and call sites; migrate any explicit catches to use the parent class.
```

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovRejectionException.java:
```diff
@@ -9,35 +9,36 @@
-    private String arc;
+import com.criticalblue.approovsdk.Approov;
 
+public class ApproovRejectionException extends TokenFetchStatusException {
```
You seem to have lost the class comment

jexh committed:
Conform to naming, simplify wrapped exceptions, comment updates

adriantuk committed:
Fix formatting and typos in exception messages and comments; Rm custom mutator

jexh reviewed (approved):
Lokks good now. Only 1 comment to update and then good to go (with traceID merged in). No need for a further review after the comment update.

adriantuk committed:
Update approov-service/src/main/java/io/approov/service/okhttp/ApproovNetworkException.java

Co-authored-by: jexh <jexh@users.noreply.github.com>
